### PR TITLE
[server] handle uncaught exception from grpc-js

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

Added a node js `uncaughtExcpetion` handler for a bug in `grpc-js`, which would try to startConnection for a channel that is no longer there (concurrency issue) and therefor throws an Error within `process.nextTick` which in turn kills node.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at b5c9115</samp>

This pull request enhances the reliability and stability of the server and ws-manager-bridge components, by adding better error handling and graceful shutdown logic for the grpc-js library. It affects the files `init.ts` and `main.ts` in these components.

</details>

## Related Issue(s)

Fixes #

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - se-handle-eb03158298</li>
	<li><b>🔗 URL</b> - <a href="https://se-handle-eb03158298.preview.gitpod-dev.com/workspaces" target="_blank">se-handle-eb03158298.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - se-handle-grpcs-js-excpetion-gha.14740</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by csweichel** [comment]
How does this code behave when the error occurs? What's the user experience?
What side effects do we expect this change to have (cc @akosyakov)?

Once we can reliably repro the issue, let's validate this change.

**Review by AlexTugarev — APPROVED** [review]

**Comment by svenefftinge** [comment]
> How does this code behave when the error occurs? What's the user experience?

The bug in grpc-js is a concurrency issue where it tries to access a channel that has already been discarded. So the user experience is not impacted.

> What side effects do we expect this change to have (cc @akosyakov)?

None

**Comment by mustard-mh** [comment]
@svenefftinge It's related to this issue https://linear.app/gitpod/issue/EXP-360#comment-576c2a99 right? Don't know if it has side affect, i.e: It can only be recovered when server restart. If it's true then we can't make workspace run anymore.

We need to be careful with this PR, and better leave preview env **untouched** until tomorrow to see what happened. (it can reproduce if we have no workspace create for a long time)

/hold

File: components/server/src/init.ts [hunk 1]
```diff
@@ -63,6 +63,24 @@ if (process.env.NODE_ENV === "development") {
...
+    const server = container.get(Server);
+    const port = 3000;
+    const app = express();
+
+    process.on("uncaughtException", function (err) {
+        // fix for https://github.com/grpc/grpc-node/blob/master/packages/grpc-js/src/load-balancer-pick-first.ts#L309
+        if (err && err.message && err.message.includes("reading 'startConnecting'")) {
+            log.error("uncaughtException", err);
+        } else {
+            throw err;
```
Comment (akosyakov) [inline components/server/src/init.ts:75]: Should we log and do process.exit(1) here? Or Node.js will do it?

Comment (svenefftinge) [reply components/server/src/init.ts:75]: node js does it, i.e. that is the same behavior as not having a handler

**Comment by svenefftinge** [comment]
We have verified this and even were lucky to see the error in the preview environment once. No user experience impact!

As said in the description, there is no need for recovery as the bug is that grpc-js tries to call `startConnecting` on a stale channel (has been cleaned up concurrently). So just logging is the correct thing.

**Comment by mustard-mh** [comment]
Confirm that it fixed hang problem 🎉

**Pull request merged**